os/fs/smartfs/smartfs_utils.c

```diff
@@ -1833,7 +2080,67 @@ int smartfs_truncatefile(struct smartfs_mountpt_s *fs, struct smartfs_entry_s *e
 
 int smartfs_shrinkfile(FAR struct smartfs_mountpt_s *fs, FAR struct smartfs_ofile_s *sf, off_t length)
 {
-	/* TODO This will be added later */
+	int ret;
+	struct smart_read_write_s readwrite;
+	struct smartfs_chain_header_s *chainheader;
+	uint16_t sector;
+	uint16_t nextsector;
+	uint8_t data_len;
+	uint8_t offset;
+
+	/* 'data_len' represents length of DATA in a sector, the chainheader is not counted in file data */
+	data_len = fs->fs_llformat.availbytes - sizeof(struct smartfs_chain_header_s);
+
+	/* Seek till point 'length' of the file */
+	smartfs_seek_internal(fs, sf, length, SEEK_SET);
+	offset = sizeof(struct smartfs_chain_header_s) + (length % data_len);
+	/* Keep as many sectors as needed and replace extra bytes in the last needed sector with ERASEDSTATE */
+#ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
+	memset(&sf->buffer[offset], CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes - offset);
+	chainheader = (struct smartfs_chain_header_s *)sf->buffer;
+	*(uint16_t *)chainheader->used = (length % data_len);
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
+
+	smartfs_sync_internal(fs, sf);
+#else
+	smartfs_setbuffer(&readwrite, sf->currsector, 0, fs->fs_llformat.availbytes, (FAR uint8_t *)fs->fs_rwbuffer);
+	ret = FS_IOCTL(fs, BIOC_READSECT, (unsigned long)&readwrite);
+	if (ret < 0) {
+		fdbg("Failed to read file sector no. %d, error = %d\n", nextsector, ret);
+		return ret;
+	}
+
+	chainheader = (struct smartfs_chain_header_s *)fs->fs_rwbuffer;
+	memset(&fs->fs_rwbuffer[offset], CONFIG_SMARTFS_ERASEDSTATE, fs->fs_llformat.availbytes - offset);
+	*(uint16_t *)chainheader->used = (length % data_len);
+	*(uint16_t *)chainheader->nextsector = SMARTFS_ERASEDSTATE_16BIT;
```

Comment:
these 2 line can be moved to above of #ifdef

Reply:
chainheader is assigned differently in both cases,
These two assignments can be made only after chainheader is assigned.

Reply:
ah yes you're right then please use one variable to reduce calculate length % data_len

Reply:
Ohkay